Fix calling nonexistent function on OfflineUser

KimTheFirst opened the pull request:
"Call to undefined method OCA\User_LDAP\User\OfflineUser::composeAndStoreDisplayName()" when accessing user list.

Same issue as #859. Same fix used.

KimTheFirst committed:
Fix calling nonexistent function on OfflineUser

"Call to undefined method OCA\User_LDAP\User\OfflineUser::composeAndStoreDisplayName()" when accessing user list.

Same issue as #859. Same fix used.

blizzz commented:
@KimTheFirst do you have a backlog of that issue?

blizzz commented on apps/user_ldap/lib/Access.php:
```diff
@@ -731,6 +731,12 @@ public function cacheUserDisplayName($ocName, $displayName, $displayName2 = '')
 		}
+		if ($user instanceof User) {
+		    $displayName = $user->composeAndStoreDisplayName($displayName, $displayName2);
+		}
```
combine per 'else if` and we're good to go

KimTheFirst committed:
Combine username cache checks into else if.


-Fix my error calling the compose function twice, which would have negated the fix.

-Add a final "else" statement with a catch-all exception for good measure, in case other unhandled user objects are added one day.

blizzz commented:
@KimTheFirst thanks, the patch looks good now. Nevertheless, I'd be interested in the backlog to ensure we don't fix just a symptom instead of underlying problem.

KimTheFirst commented:
@blizzz This is what I was getting in the log file: https://pastebin.com/TMnHQGr0

NC is also saying that the user does not exist anymore (They do, they are in LDAP). There is definitely a deeper error here but it might be elsewhere in the infrastructure. In the meantime, I needed this patch to regain access to the users section.

I will update you here when I've gotten to the bottom of it.

KimTheFirst commented:
It does indeed look similar, but possibly with a different root cause. In both cases user lookup errors cause the users section to get blocked.

KimTheFirst commented:
@spanguel 

I can't tell you if it's safe to delete them because I don't know if those users have data in your system or what your specific circumstances are. Iif you are asking whether or not deleting these users will damage your NC installation then the answer is that you can freely delete any user you want and your service should remain operational.

As for the error message being the same - this patch is to do with the way user lookups are being handled rather than any specific error. When a user is considered "offline" by nextcloud then it can cause this error, regardless of the specific reason for that user being offline.

blizzz commented:
#9502 confirmed it is not because the apostrophe, but as here, when the user was marked deleted. Therefore, i'm closing this as a dup.

blizzz commented:
forgot it is a PR, sorry, still looking into as the underlying behaviour is plain wrong.

blizzz commented:
The fix for #9502 (dup, most likely) was released with 13.0.3 → fixed. There we no reports since either.
